support/controlplane-component/README.md

```diff
@@ -0,0 +1,278 @@
+
+const (
+    // This is the name of your manifests directory.
+    ComponentName = "my-component"
+)
+
+func NewComponent() component.ControlPlaneComponent {
+	return component.NewDeploymentComponent(ComponentName, &MyComponent{}).
+		Build()
+}
+```
+
+!!! note 
+
+    use `component.NewStatefulSetComponent()` instead if your components's workload is a StatefulSet.
+
+
+## Registering your component
+
+Finally your component needs to be be registered in `/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go` inside `registerComponents()` method
+
+```go
+// control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go
+
+import (
+    mycomponentv2 "github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/v2/mycomponent"
+)
+
+func (r *HostedControlPlaneReconciler) registerComponents() {
+	r.components = append(r.components,
+        // other components  
+        ...
+		mycomponentv2.NewComponent(),
+	)
+}
+```
+
+## Customizing your component
+
+Some components require dynamic additional config, arguments, env variables, etc. based on the `HostedControlPlane` spec. You can adapt your manifests dynamically by defining adapt functions with the following signature `func(ControlPlaneContext, resourceType) error`
+
+To adapt your deployment, first define your adapt function:
+```go
+// control-plane-operator/controllers/hostedcontrolplane/v2/mycomponent/component.go
+
+import (
+    "github.com/openshift/hypershift/support/util"
+    component "github.com/openshift/hypershift/support/controlplane-component"
+)
+
+func adaptDeployment(cpContext component.ControlPlaneContext, deployment *appsv1.Deployment) error {
+    // for example, append a new arg to a container
+    util.UpdateContainer("container-name", deployment.Spec.Template.Spec.Containers, func(c *corev1.Container) {
+		c.Args = append(c.Args,
+			"--platform-type", string(cpContext.HCP.Spec.Platform.Type),
+		)
+    })
+}
+```
+
+Then register your adapt function when creating the `ControlPlaneComponent` instance
+
+```go hl_lines="5"
+// control-plane-operator/controllers/hostedcontrolplane/v2/mycomponent/component.go
+
+func NewComponent() component.ControlPlaneComponent {
+	return component.NewDeploymentComponent(ComponentName, &MyComponent{}).
+        WithAdaptFunction(adaptDeployment).
+		Build()
+}
+```
+
+Similarly to adapt any other manifest, define your adapt function then register it using `.WithManifestAdapter("manifest-name.yaml", component.WithAdaptFunction(func))`:
+
+```go hl_lines="6 7 8 9"
+// control-plane-operator/controllers/hostedcontrolplane/v2/mycomponent/component.go
+
+func NewComponent() component.ControlPlaneComponent {
+	return component.NewDeploymentComponent(ComponentName, &MyComponent{}).
+        WithAdaptFunction(adaptDeployment).
+        WithManifestAdapter(
+			"my-config.yaml",
+			component.WithAdaptFunction(adaptMyConfig),
```

Comment:
is it necessary to cast to the WithAdaptFunction type?

Reply:
WithAdaptFunction dosn't expect any specific type, it accepts any function with the following signature `[T client.Object] func(cpContext ControlPlaneContext, resource T) error` 
It is the responsibility of the components' developer to pass a function with the correct type.